custom_components/ha_washdata/ml/trainer.py

```
@@ -0,0 +1,413 @@
+plus a weight vector, bias, and decision threshold - which the lab fits with a
+short pure-NumPy gradient descent (``wash_ml/end_detection.py::_fit_logistic``).
+This module reproduces that fit and the exact scoring math the embedded
+``*_model.py`` modules use, so a model trained here on the user's own cycles is
+byte-compatible with the shipped baseline and can be scored identically.
+
+No new dependencies: NumPy only. Nothing here runs unless the caller (behind the
+``ENABLE_ML_TRAINING`` flag) invokes it.
+"""
+from __future__ import annotations
+
+from typing import Any, Mapping, Sequence
```

Ruff (0.15.20)

[warning] Import from `collections.abc` instead: `Mapping`, `Sequence`

Import from `collections.abc`

(UP035)